### ISSUE TITLE
Improve error messages

### ISSUE THREAD

**Issue by NathanReb** [ocaml-ppx/ppx_deriving_yojson#80]
It would be nice to improve the error reporting of the derived `of_yojson` functions as it would really ease debugging errors in the parsed json or in hand written parsers involved.

Some of them can be improved quite easily:
```ocaml
[%of_yojson: int] `Null
```
is
```ocaml
Error ""
```

I think in general it would also be a huge improvement to be able to track the error location through the json structure.

For example, if we consider the following:
```ocaml
type t =
  { a : A.t
  ; b : B.t
  }
  [@@deriving of_yojson]
```
it would be helpful to concatenate `"in field a: "` to the `A.of_yojson` error message when the parsing of `a` fails.

I don't have a clear idea of what would be the best way to handle those but would be happy to discuss it with you and to provide a PR if you feel like that should be improved.

**Comment by gasche** [comment]
The requirement/need/wish seems very sensible to me. Two remarks:

- The error-reporting scheme should not hurt performance too much. I am highly confident that it's possible to have both at the same time (your idea of recursively passing an "access path" to the parser to report where the problem is seems mostly free), but I would expect any change to come with summary benchmarks to measure this.

- ppx_deriving_yojson is supposed to desugar into straightforward calls to the `Yojson` library, not implement elaborate logic itself. Ideally we want all "interesting" logic to be implemented at the Yojson level, and simply used from the deriver. In the present case, however, I am not completely sure how to do this: error-reporting may require transporting information as we traverse data structures, so we need to do something on our end. I would still expect most of the logic to be added/improved at the Yojson level.

**Comment by whitequark** [comment]
(Agreed on both counts.)

**Comment by emillon** [comment]
Just improving error messages in these case (without adding location) is a nice quick win. I just got bit by `[%of_yojson: string]` returning `Error ""` and had no idea where to start :)

**Comment by NathanReb** [comment]
I am reviving this issue as I recently got bit by quite unhelpful errors when parsing JSON objects. I think there a few low hanging fruits to improve the quality of Objects parsing errors:
1. When the actual JSON is not an object, the current error message is simply the type name, it could also mention that a JSON object is expected.
2. When the parsers encounters an unsupported key, the error message is simply the type name, again it should probably mention the key name and that it is unsupported.
3. Missing mandatory fields are simply reported by their name, adding the mention "missing key" would go a long way I think.

I think those should be fairly easy to add as they require no change in the parsing code (except for binding the unknown keys in the pattern matching).

Another slightly more complex change could be to aggregate errors, e.g. if there are multiple missing fields in an object, only the first one will be reported even though we could report them all.

Overall I think it would be nice to provide an alternate error type to the current raw string that users of ppx_deriving_yojson generated parsers could interpret and turn into end-user-friendly error messages but this is a bit more involved as well and I think the three suggestions above would already fix quite a lot of issues as they can be understood by end users.

I'm happy to have a go at this and open a PR!

**Comment by gasche** [comment]
Yep, I think that attacking the low-hanging fruits and improving error messages gradually is a good strategy!
